File: backend/internal/repository/account.go
package repository

import (
	"errors"

	"github.com/sky/personal-ledger/internal/model"
	"gorm.io/gorm"
)

var ErrAccountBalancePreventsDeletion = errors.New("account balance prevents deletion")
var ErrUnsafeAccountFieldUpdate = errors.New("unsafe account field update")

var accountMetadataFields = map[string]struct{}{
	"name":          {},
	"icon":          {},
	"color":         {},
	"payment_day":   {},
	"billing_day":   {},
	"credit_limit":  {},
	"interest_rate": {},
	"start_date":    {},
	"target_date":   {},
	"remark":        {},
	"is_archived":   {},
}

type AccountRepository struct {
	db *gorm.DB
}

func NewAccountRepository(db *gorm.DB) *AccountRepository {
	return &AccountRepository{db: db}
}

func (r *AccountRepository) Create(account *model.Account) error {
	return r.db.Create(account).Error
}

func (r *AccountRepository) GetByID(id string) (*model.Account, error) {
	var account model.Account
	err := r.db.First(&account, "id = ?", id).Error
	if err != nil {
		return nil, err
	}
	return &account, nil
}

func (r *AccountRepository) GetByUserID(userID uint, includeArchived bool) ([]model.Account, error) {
	var accounts []model.Account
	query := r.db.Where("user_id = ?", userID)
	if !includeArchived {
		query = query.Where("is_archived = ?", false)
	}
	err := query.Order("sort_order ASC, created_at ASC").Find(&accounts).Error
	return accounts, err
}

func (r *AccountRepository) GetByUserIDForHistory(userID uint) ([]model.Account, error) {
	var accounts []model.Account
	err := r.db.Unscoped().Where("user_id = ?", userID).
		Order("created_at ASC, id ASC").
		Find(&accounts).Error
	return accounts, err
}

func (r *AccountRepository) UpdateMetadataForUser(id string, userID uint, updates map[string]any) error {
	for field := range updates {
		if _, ok := accountMetadataFields[field]; !ok {
			return ErrUnsafeAccountFieldUpdate
		}
	}
	result := r.db.Model(&model.Account{}).
		Where("id = ? AND user_id = ?", id, userID).
		Updates(updates)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected != 1 {
		var count int64
		if err := r.db.Model(&model.Account{}).Where("id = ? AND user_id = ?", id, userID).Count(&count).Error; err != nil {
			return err
		}
		if count == 0 {
			return gorm.ErrRecordNotFound
		}
	}
	return nil
}

func (r *AccountRepository) DeleteForUserIfBalanceAllows(id string, userID uint) error {
	result := r.db.
		Where("id = ? AND user_id = ? AND (is_archived = ? OR current_balance = ?)", id, userID, true, 0).
		Delete(&model.Account{})
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 1 {
		return nil
	}

	var account model.Account
	if err := r.db.Select("id", "current_balance", "is_archived").
		First(&account, "id = ? AND user_id = ?", id, userID).Error; err != nil {
		return err
	}
	if !account.IsArchived && account.CurrentBalance != 0 {
		return ErrAccountBalancePreventsDeletion
	}
	return gorm.ErrRecordNotFound
}

func (r *AccountRepository) DeleteAllByUserID(userID uint) error {
	return r.db.Where("user_id = ?", userID).Delete(&model.Account{}).Error
}

func (r *AccountRepository) UpdateSortOrder(ids []string) error {
	return r.db.Transaction(func(tx *gorm.DB) error {
		for i, id := range ids {
			if err := tx.Model(&model.Account{}).Where("id = ?", id).
				Update("sort_order", i).Error; err != nil {
				return err
			}
		}
		return nil
	})
}

func (r *AccountRepository) CreateBatch(accounts []model.Account) error {
	return r.db.Create(&accounts).Error
}

func (r *AccountRepository) DB() *gorm.DB {
	return r.db
}

File: web/src/api/account.ts
import { get, post, put, del, patch } from '@/utils/request'
import type { AccountType } from '@/utils/accountTypeRules'

export type { AccountType } from '@/utils/accountTypeRules'
export { DEBT_ACCOUNT_TYPES, isDebtAccount } from '@/utils/accountTypeRules'

export interface Account {
  id: string
  name: string
  type: AccountType
  icon: string
  color: string
  initial_balance: number
  current_balance: number
  payment_day?: number
  billing_day?: number
  credit_limit?: number
  interest_rate?: number
  total_paid: number
  start_date?: string
  target_date?: string
  paid_off_at?: string
  remark?: string
  is_archived: boolean
  sort_order: number
}

export interface AccountListResponse {
  list: Account[]
  total_assets: number
  total_liabilities: number
  net_assets: number
}

export interface CreateAccountParams {
  name: string
  type: AccountType
  icon?: string
  color?: string
  initial_balance: number
  payment_day?: number
  billing_day?: number
  credit_limit?: number
  interest_rate?: number
  start_date?: string
  target_date?: string
  remark?: string
}

export interface UpdateAccountParams {
  name?: string
  icon?: string
  color?: string
	payment_day?: number | null
	billing_day?: number | null
	credit_limit?: number | null
	interest_rate?: number | null
	start_date?: string | null
	target_date?: string | null
	remark?: string | null
}

export const accountApi = {
  getList(includeArchived = false): Promise<AccountListResponse> {
    return get<AccountListResponse>('/accounts', {
      params: { include_archived: includeArchived }
    })
  },

  getById(id: string): Promise<Account> {
    return get<Account>(`/accounts/${id}`)
  },

  create(params: CreateAccountParams): Promise<Account> {
    return post<Account>('/accounts', params)
  },

	update(id: string, params: UpdateAccountParams): Promise<Account> {
		return patch<Account>(`/accounts/${id}`, params)
  },

  delete(id: string): Promise<void> {
    return del<void>(`/accounts/${id}`)
  },

  archive(id: string, isArchived: boolean): Promise<void> {
    return patch<void>(`/accounts/${id}/archive`, { is_archived: isArchived })
  },

  updateSort(ids: string[]): Promise<void> {
    return put<void>('/accounts/sort', { ids })
  },
}
